### PR TITLE
adding tests and defaulting, and validation to the api

### PR DESCRIPTION
**Pull request by n3wscott**

### CODE REVIEW

File: pkg/apis/discovery/v1alpha1/ducktype_defaults.go [hunk 1]
```diff
@@ -18,9 +18,20 @@ package v1alpha1
 
 import (
 	"context"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_defaults.go:20]: Format Go code:
```suggestion
	"context"
	"strings"

```

File: pkg/apis/discovery/v1alpha1/ducktype_defaults.go [hunk 2]
```diff
@@ -18,9 +18,20 @@ package v1alpha1
...
+	"knative.dev/pkg/apis"
+	"strings"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_defaults.go:22]: Format Go code:
```suggestion

```

File: pkg/apis/discovery/v1alpha1/ducktype_defaults_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1alpha1
+
+import (
+	"context"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_defaults_test.go:19]: Format Go code:
```suggestion
	"context"
	"testing"

```

File: pkg/apis/discovery/v1alpha1/ducktype_defaults_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/google/go-cmp/cmp"
+	"testing"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_defaults_test.go:21]: Format Go code:
```suggestion

```

File: pkg/apis/discovery/v1alpha1/ducktype_lifecycle_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1alpha1
+
+import (
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_lifecycle_test.go:18]: Format Go code:
```suggestion
import (
	"sort"
	"testing"

```

File: pkg/apis/discovery/v1alpha1/ducktype_lifecycle_test.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	"sort"
+	"testing"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_lifecycle_test.go:24]: Format Go code:
```suggestion

```

File: pkg/apis/discovery/v1alpha1/ducktype_validation.go [hunk 7]
```diff
@@ -18,16 +18,93 @@ package v1alpha1
 
 import (
 	"context"
+	"fmt"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_validation.go:21]: Format Go code:
```suggestion
	"fmt"
	"strings"

```

File: pkg/apis/discovery/v1alpha1/ducktype_validation.go [hunk 8]
```diff
@@ -18,16 +18,93 @@ package v1alpha1
...
+	"k8s.io/apimachinery/pkg/labels"
+	"strings"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_validation.go:23]: Format Go code:
```suggestion

```

File: pkg/apis/discovery/v1alpha1/ducktype_validation_test.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1alpha1
+
+import (
+	"context"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_validation_test.go:19]: Format Go code:
```suggestion
	"context"
	"testing"

```

File: pkg/apis/discovery/v1alpha1/ducktype_validation_test.go [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	"testing"
```
Comment (mattmoor) [inline pkg/apis/discovery/v1alpha1/ducktype_validation_test.go:23]: Format Go code:
```suggestion

```

File: pkg/apis/discovery/v1alpha1/ducktype_validation_test.go [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+					Names: DuckTypeNames{
+						Name:     "ThisDuck",
+						Plural:   "thisducks",
+						Singular: "thisduck",
+					},
+					Versions: []DuckVersion{{
+						Name: "v1",
+					}},
+					Selectors: []CustomResourceDefinitionSelector{{
+						LabelSelector: "turn down for what",
```
Comment (lberk) [inline pkg/apis/discovery/v1alpha1/ducktype_validation_test.go:257]: :rofl:

**Comment by pierDipi** [comment]
FYI pre-submit jobs aren't running tests: https://github.com/knative-sandbox/discovery/blob/master/test/presubmit-tests.sh

**Pull request merged**